File: 0378-Kth Smallest Element in a Sorted Matrix/0378-Kth Smallest Element in a Sorted Matrix.cpp
class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
        int n = matrix.size();
        priority_queue<Cell, vector<Cell>, cmp> pq;
        pq.emplace(0, 0, matrix[0][0]);
        int result;
        while (k--)
        {
            Cell cell = pq.top();
            pq.pop();
            result = cell.val;
            int row = cell.row;
            int col = cell.col;
            if (row < n - 1)
            {
                pq.emplace(row + 1, col, matrix[row + 1][col]);
            }
            if (row == 0 && col < n - 1)
            {
                pq.emplace(row, col + 1, matrix[row][col + 1]);
            }
        }
        
        return result;
    }
    
private:
    struct Cell {
        int row;
        int col;
        int val;
        Cell (int r, int c, int v): row(r), col(c), val(v){}
    };
    
    struct cmp {
        bool operator() (Cell& c1, Cell& c2) {
            return c1.val > c2.val;
        }
    };
};

// Range binary search, O(nlog(max - min))
class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
        int n = matrix.size();
        int low = matrix[0][0], high = matrix[n - 1][n - 1];
        while (low < high)
        {
            int mid = low + (high - low) / 2;
            int count = 0;
            int row = n - 1, col = 0;
            while (row >= 0 && col < n)
            {
                if (matrix[row][col] > mid)
                {
                    --row;
                }
                else
                {
                    ++col;
                    count += row + 1;
                }
            }
            
            if (count < k)
            {
                low = mid + 1;
            }
            else
            {
                high = mid;
            }
        }
        
        return low;
    }
};

File: 0301-0400/0378-Kth Smallest Element in a Sorted Matrix/0378-Kth Smallest Element in a Sorted Matrix.py
import heapq
class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        pq = [(matrix[0][0], 0, 0)]
        while k > 0:
            result, row, col = heapq.heappop(pq)
            if row < len(matrix) - 1:
                heapq.heappush(pq, (matrix[row + 1][col], row + 1, col))
            if row == 0 and col < len(matrix[0]) - 1:
                heapq.heappush(pq, (matrix[row][col + 1], row, col + 1))
            k -= 1
        return result


class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        low, high = matrix[0][0], matrix[-1][-1]
        while low < high:
            mid = (low + high) // 2
            count = col = 0
            row = len(matrix) - 1
            while row >= 0 and col < len(matrix[0]):
                if matrix[row][col] > mid:
                    row -= 1
                else:
                    count += row + 1
                    col += 1
            if count < k:
                low = mid + 1
            else:
                high = mid
        return low
